halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/config/v1/HalconfigParser.java

```diff
@@ -141,6 +158,42 @@ public void saveConfig() {
     saveConfigTo(Paths.get(halconfigPath));
   }
 
+  /**
+   * Deletes all files in the staging directory that are not referenced in the hal config.
+   */
+  public void cleanLocalFiles(Path stagingDirectoryPath) {
+    if (GlobalApplicationOptions.getInstance().isUseRemoteDaemon()) {
+      Halconfig halconfig = getHalconfig();
+      Set<String> referencedFiles = new HashSet<String>();
+      Consumer<Node> fileFinder = n -> referencedFiles.addAll(n.localFiles().stream().map(f -> {
+        try {
+          f.setAccessible(true);
+          return (String) f.get(n);
+        } catch (IllegalAccessException e) {
+          throw new HalException(FATAL, "Failed to clean staging directory: " + e.getMessage(), e);
```

Comment:
This feels more like a runtime error than a halexception - by convention the latter is reserved for errors the user can do something about.

Reply:
SGTM.